source/compiler/qsc_frontend/src/resolve/imports.rs

```diff
@@ -0,0 +1,524 @@
+                    || namespace.items.iter().map(AsRef::as_ref),
+                    attempted_imports,
+                ) {
+                    new_imports_were_added = true;
+                }
+
+                resolver.pop_scope();
+            }
+        }
+
+        new_imports_were_added
+    });
+    resolver.errors.extend(errors);
+}
+
+/// Resolve all imports declared in top-level statements.
+/// Imports are then made available in the persistent local scope.
+pub(super) fn resolve_top_level_imports(resolver: &mut Resolver, package: &Package) {
+    let errors = iterate_until_done(|attempted_imports| {
+        try_resolve_imports_and_exports(
+            resolver,
+            || {
+                package.nodes.iter().filter_map(|node| {
+                    if let TopLevelNode::Stmt(stmt) = node {
+                        if let StmtKind::Item(item) = &*stmt.kind {
+                            return Some(item.as_ref());
+                        }
+                    }
+                    None
+                })
+            },
+            attempted_imports,
+        )
+    });
+    resolver.errors.extend(errors);
+}
+
+/// Resolve all imports declared in a block.
+/// Imports are then made available in that block's scope.
+pub(super) fn resolve_block_imports(resolver: &mut Resolver, block: &Block) {
+    let errors = iterate_until_done(|attempted_imports| {
+        try_resolve_imports_and_exports(
+            resolver,
+            || {
+                block.stmts.iter().filter_map(|stmt| {
+                    if let StmtKind::Item(item) = &*stmt.kind {
+                        Some(item.as_ref())
+                    } else {
+                        None
+                    }
+                })
+            },
+            attempted_imports,
+        )
+    });
+    resolver.errors.extend(errors);
+}
+
+/// Makes a single pass over the imports and exports in the current scope,
+/// attempting to resolve them.
+///
+/// Once an import has been successfully resolved, its name is made available
+/// in the current scope.
+///
+/// Exports are only handled if the current scope is a namespace, and
+/// once successfully resolved, their names will be added to the global scope.
+///
+/// Note that this single pass may not result in all imports being resolved.
+/// Multiple passes will ensure that any imports referencing other imports
+/// in the same scope are resolved.
+///
+/// For this reason, resolution errors are not reported immediately, but
+/// instead stored in `attempted_imports` to be retried later. Returns
+/// `true` if any new imports or exports were bound in the current scope,
+/// as a signal that we should continue iterating.
+fn try_resolve_imports_and_exports<'a, I>(
+    resolver: &mut Resolver,
+    scope_items_iter: impl Fn() -> I,
+    attempted_imports: &mut FxHashMap<NodeId, Result<(), Error>>,
+) -> bool
+where
+    I: Iterator<Item = &'a Item>,
+{
+    if scope_items_iter().count() == 0 {
```

Comment:
Assuming we need separate ordered loops here, I think we should collect the iterator once into a vec and update the usages for the vec.